independent-projects/bootstrap/runner/src/main/java/io/quarkus/bootstrap/runner/RunnerClassLoader.java

```diff
@@ -96,19 +111,24 @@ public Class<?> loadClass(String name, boolean resolve) throws ClassNotFoundExce
             String classResource = name.replace('.', '/') + ".class";
             for (ClassLoadingResource resource : resources) {
                 accessingResource(resource);
-                byte[] data = resource.getResourceData(classResource);
+                ByteBuffer pooledData = this.pooledData;
+                this.pooledData = null;
```

Comment:
This method can be called concurrently, so there probably should be some kind of atomicity or mutual exclusion or something here if exclusive ownership is required.

Reply:
Ah didn't knew...
@geoand this is supposed to be called by a singe thread each time? because effectively I can see many read locks/synchronized all around which sound suspicious, including the comment from @dmlloyd

Reply:
With class loaders, we just don't have control over the callers.

Reply:
In the initial phase, all classl loading should be done on the main thread, but there is no way we can guarantee this unfortunately.

Reply:
> but there is no way we can guarantee this unfortunately.

I can decide to enforce it by checking which thread is asking to load the class: if it's the main thread AND we're in the startup I can use such optimization; just need a way to communicate it

Reply:
I think that is worth benchmarking

Reply:
I've added a new commit which "address" (by cheating? We need test coverage for something like that!) it

Reply:
The downside of such is that any Thread which is still within the "startup" time window, but cause class loading, won't be able to benefit of any pooling: one example of this is https://github.com/quarkusio/quarkus/blob/f032a0bdba4314fdd9f5dfb3a820db84b63201de/extensions/netty/runtime/src/main/java/io/quarkus/netty/runtime/NettyRecorder.java#L20 and `NioEventLoop::run` 

see, in the allocation flamegraph

![image](https://github.com/quarkusio/quarkus/assets/13125299/1b91f114-6c1f-4aa0-a0be-e3cc22b29725)

which won't benefit from such pooling: in the original PR it was allocating 5.7 MB, while with this last changes 6.7 MB, which is exactly the same as before any of these changes.

In short:
- using main thread id confinement: 34.5 MB allocated
- without main thread id confinement, still pooling: 33.4 MB allocated
- main: 38.9 MB

Which still means a relevant saving IMO, but need to understand how it get worsen/better with hinernate in the mix, which will be my next test.

Reply:
That's very interesting.

For cases like the one you mention, we can likely have the corresponding extension force class loading of the specific classes in the main thread

Reply:
Exactly, and we will save a lot of heap, cause it's a safe state which can optimize more.
I have no idea (yet, I am still learning how it works), what's the degree of allowed parallelism of the class loader, but if it's not allowing concurrent I/O for different unrelated classes, it will be a win.
Conversely, if it allow that level of parallelism, maybe worth leveraging it, but being aware that:
- dispatching class loading to other native threads is still costy both in memory and handoff time 
- every extension should have ALOT of I/O to be performed, or the cost of synchronization on class definition (and contention) will negate it
-

Reply:
What I am proposing essentially to mitigate thew Netty case you mentioned above would be something like:

```java
    public void eagerlyInitChannelId() {
        Class.forName("whatever.netty.classes.we.need", false, TCCL);
        new Thread(new Runnable() {
            @Override
            public void run() {
                ...
            }
        }).start();
    }
```

Obviouysly that could be made more general if necessary

Reply:
Yes, this is an easy win for this specific use case, I can tell: let me give it a shot!